### ISSUE TITLE
Unmet Dependencies error on latest docker images

### ISSUE THREAD

**Issue by varunagrawal** [osrf/docker_images#196]
On using the `osrf/ros:indigo-desktop-full` image about 12 days ago, I was easily able to build a docker file on top of this image and install the fetch gazebo package.

However, today when I try to build the container from scratch, I get the following error:

```
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 ros-indigo-fetch-gazebo-demo : Depends: ros-indigo-fetch-moveit-config but it is not going to be installed
                                Depends: ros-indigo-moveit-python but it is not going to be installed
                                Depends: ros-indigo-simple-grasping but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
``` 

The Docker command I am running is `RUN apt-get update && apt-get -f install -y ros-indigo-fetch-gazebo-demo`.

Any ideas what may be the issue? I am guessing something wrong in the images themselves since as per dockerhub, they were updated 2 days ago.

**Comment by ruffsl** [comment]
This doesn't seem to be docker related, but more of a packaging issue. I just build a freatch `ros:indigo-ros-core-trusty` image locally and still encounter the above 'broken packages' message. At first I attempted to just trigger the docker build of the desktop images, as nesting images still an issue sometime with respect to freshness of the layers in the nested tags, but this doesn't seem to be the case here. I'd say bringing up the issue with the package [maintainer for fetch-gazebo-demo](https://github.com/fetchrobotics/fetch_gazebo/blob/eecb032ded6f675ded81e477087218a2079944fa/fetch_gazebo_demo/package.xml#L8). On a side note, `ros-indigo-fetch-gazebo` still seems installable:

<details>

``` shell
$ docker run -it --rm osrf/ros:indigo-desktop-full 
root@2ee3679eb76e:/# apt-get update && apt-get -f install -y ros-indigo-fetch-gazebo
Get:1 http://packages.ros.org trusty InRelease [4031 B]
Get:2 http://packages.ros.org trusty/main amd64 Packages [873 kB]              
Get:3 http://security.ubuntu.com trusty-security InRelease [65.9 kB]        
Ign http://archive.ubuntu.com trusty InRelease                                 
Get:4 http://security.ubuntu.com trusty-security/universe Sources [101 kB]
Get:5 http://archive.ubuntu.com trusty-updates InRelease [65.9 kB]
Get:6 http://security.ubuntu.com trusty-security/main amd64 Packages [954 kB]
Get:7 http://archive.ubuntu.com trusty-backports InRelease [65.9 kB]           
Get:8 http://archive.ubuntu.com trusty Release.gpg [933 B]                     
Get:9 http://archive.ubuntu.com trusty-updates/universe Sources [265 kB]
Get:10 http://security.ubuntu.com trusty-security/restricted amd64 Packages [18.1 kB]
Get:11 http://security.ubuntu.com trusty-security/universe amd64 Packages [325 kB]
Get:12 http://security.ubuntu.com trusty-security/multiverse amd64 Packages [4726 B]
Get:13 http://archive.ubuntu.com trusty-updates/main amd64 Packages [1376 kB] 
Get:14 http://archive.ubuntu.com trusty-updates/restricted amd64 Packages [21.4 kB]
Get:15 http://archive.ubuntu.com trusty-updates/universe amd64 Packages [620 kB]
Get:16 http://archive.ubuntu.com trusty-updates/multiverse amd64 Packages [16.0 kB]
Get:17 http://archive.ubuntu.com trusty-backports/main amd64 Packages [14.7 kB]
Get:18 http://archive.ubuntu.com trusty-backports/restricted amd64 Packages [40 B]
Get:19 http://archive.ubuntu.com trusty-backports/universe amd64 Packages [52.5 kB]
Get:20 http://archive.ubuntu.com trusty-backports/multiverse amd64 Packages [1392 B]
Get:21 http://archive.ubuntu.com trusty Release [58.5 kB]
Get:22 http://archive.ubuntu.com trusty/universe Sources [7926 kB]
Get:23 http://archive.ubuntu.com trusty/main amd64 Packages [1743 kB]
Get:24 http://archive.ubuntu.com trusty/restricted amd64 Packages [16.0 kB]
Get:25 http://archive.ubuntu.com trusty/universe amd64 Packages [7589 kB]
Get:26 http://archive.ubuntu.com trusty/multiverse amd64 Packages [169 kB]     
Fetched 22.3 MB in 6s (3255 kB/s)                                              
Reading package lists... Done
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following extra packages will be installed:
  ros-indigo-control-toolbox ros-indigo-fetch-description
  ros-indigo-realtime-tools ros-indigo-rgbd-launch
  ros-indigo-robot-controllers ros-indigo-robot-controllers-interface
  ros-indigo-robot-controllers-msgs
The following NEW packages will be installed:
  ros-indigo-control-toolbox ros-indigo-fetch-description
  ros-indigo-fetch-gazebo ros-indigo-realtime-tools ros-indigo-rgbd-launch
  ros-indigo-robot-controllers ros-indigo-robot-controllers-interface
  ros-indigo-robot-controllers-msgs
0 upgraded, 8 newly installed, 0 to remove and 6 not upgraded.
Need to get 7645 kB of archives.
After this operation, 27.4 MB of additional disk space will be used.
Get:1 http://packages.ros.org/ros/ubuntu/ trusty/main ros-indigo-realtime-tools amd64 1.9.1-0trusty-20170313-091454-0700 [25.4 kB]
...
```

**Comment by mikaelarguedas** [comment]
I believe this is not related to the docker images or the fetch-gazebo* packages as I can reproduce it with only a trusty base image and no specific ROS packages.

Looking at the failing dependency chain, it looks like the version of `python-pyassimp` in the ROS repos is not installable causing any downstream package to fail to install as well.

- Without the ROS sources, installing pyassimp works:

```
$ docker run -it --rm ubuntu:trusty
# apt update
# apt install python-pyassimp
```

- With the ROS sources:
```
$ docker run -it --rm ubuntu:trusty
# apt-get update && apt-get install -q -y     dirmngr     gnupg2     lsb-release     && rm -rf /var/lib/apt/lists/*
# apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 421C365BD9FF1F717815A3895523BAEEB01FA116
# echo "deb http://packages.ros.org/ros/ubuntu `lsb_release -sc` main" > /etc/apt/sources.list.d/ros-latest.list
# apt update && apt install python-pyassimp
The following packages have unmet dependencies:
 python-pyassimp : Depends: libassimp3 (>= 3.0~dfsg-3) but 3.0~dfsg-2 is to be installed
E: Unable to correct problems, you have held broken packages.
```

It looks like the ROS repos provide a custom [python-pyassimp](http://repositories.ros.org/ubuntu/main/pool/main/a/assimp/) for all architectures but a custom libassimp3 only for armhf so it fails on other arches. I didn't find the source or rationale for the import and shipping of a custom pyassimp.

Maybe @tfoote has more context about these custom packages and if it's prefertred to just track upstream for all arches of rebuild the custom packages to use compatible version requirements with upstream.

**Comment by varunagrawal** [comment]
Interesting. I encountered this on the docker image so I assumed this would be a good starting point, but I am confused as to why the docker image does not install the correct version of `python-pyassimp`. Or is this a result of some upstream change?

**Comment by mikaelarguedas** [comment]
> Or is this a result of some upstream change?

It's unclear to me why it started failing now (as this custom pyassimp package has been in the ROS repos since 2014). Though I think you would have the same issue on a native Trusty system and not related to the docker images.

The failure may be due to an upstream change of version of `libassimp3` (dependency of `python-pyassimp`) that makes the packages hosted on the ROS repos not installable anymore. But we would need to track down new versions upstream as well as the need for a custom package to find the proper way to fix the current installation issue

**Comment by varunagrawal** [comment]
Thanks for the input @mikaelarguedas.

I did some more digging around and found that `python-pyassimp` depends on `libassimp3` version `3.0~dfsg-3` but Ubuntu 14.04 only has upto `3.0~dfsg-2`. This is strange since the [Ubuntu package page](https://packages.ubuntu.com/trusty/python/python-pyassimp) lists `3.0~dfsg-2` as the dependency. My guess is this is an issue from `python-pyassimp`.

**Comment by mikaelarguedas** [comment]
Looks like the only change between the 2 versions is:
```
assimp (3.0~dfsg-3) unstable; urgency=low

  * Inline IOSystem::ComparePaths

 -- Austin Hendrix <EMAIL_ADDRESS>  Thu, 14 Aug 2014 09:38:05 +0000
```

@ahendrix do you happen to remember why this change was necessary and if it needs to be shiped on all architectures?

**Comment by trainman419** [comment]
It's been 4 years. I have some vague memory that this prevented either a crash or a multiply defined symbol error.

I don't think it would hurt to apply the fix on x86, but at the time I was only working on armhf.

**Comment by tfoote** [comment]
I think that this might be a consequence of our recent switch of repo hosting for the boostrap repository.

https://discourse.ros.org/t/planned-downtime-for-ros-bootstrap-repository/6018

 One of the ramifications of the update bootstrap repository hosting is that architecture all packages automatically flooded to all architectures where previously they might have only been indexed for one. To that end I believe that @trainman419 packaged it for armhf, but with the recent switchover that custom packaged version might now be in use on all platforms.

@nuclearsandwich 

Browsing the import_upstream job I don't see this file being imported but I think this might be why things changed recently.

**Comment by nuclearsandwich** [comment]
> I think that this might be a consequence of our recent switch of repo hosting for the boostrap repository.

Both the timeline and symptoms line up with the bootstrap repo migration. It is weird that it doesn't show up in the logs and doubly weird that the package is indexed but not actually available. Doing a bit of testing and then trying to hunt down the answers to some of the anomalies.

**Comment by nuclearsandwich** [comment]
The inclusion is in the dry-run output: http://build.ros.org/job/import_upstream/269/consoleText

```
aptmethod got 'http://repos.ros.org/repos/ros_bootstrap/dists/trusty/Release'
Updates needed for 'trusty|main|amd64':
...
add 'libassimp-doc' - '3.0~dfsg-3' 'ros_bootstrap'
...
add 'python-pyassimp' - '3.0~dfsg-3' 'ros_bootstrap'
...
```
Aptly does not as I understand it have a way to list `Architecture: all` packages in only certain architectures. In addition to building libassimp for other architectures, we could also rebuild or modify our python-pyassimp and libassimp-doc packages to be `Architecture: armhf` so they won't be added to other architectures by aptly.

**Comment by nuclearsandwich** [comment]
I've updated the python-pyassimp (and am just now realizing I forgot to do libassimp-doc) to target only the armhf architecture and propagated that change through our repositories. This should allow architectures other than armhf to install correctly now.

However when I modified the armhf package I didn't take into account the dpkg format differences between trusty and bionic so the armhf package isn't installable. I hope to remedy this later this evening and publish fixed and installable packages for python-pyassimp and libassimp-doc for armhf. And then purge the libassimp-doc package from our bootstrap repositories for other architectures as I have already done for python-pyassimp.
